File: src/readBoxcal.m
function boxcal=readBoxcal(fpath, fname, nc)
% readBoxcal reads an ascii instrument response file of the legacy Phoenix 
% format (MTU-5A) and output the "boxcal" frequency-response cell array
% fpath: path to the calibration file 
% fname: name of the calibration file (exluding extensions)
%
% boxcal: output cell array of the instrument response

% check the input arguments
if nargin < 3
    % 5 for MTU-5 
    nc = 5;
end
% start reading calibration files
if exist([fpath fname],'file')
    disp(['opening box calibration file: ',fname]);
else
    errordlg(['Box calibration ' fname ' not found','?']);
    return;
end
% set calibration length here (hard coded)
% just play safe - normally it should be of length 99
calength=150;
% allocate box cell array
boxcal=cell(nc,1);
for i=1:nc
    boxcal{i}.channel=i;
    boxcal{i}.freq=zeros(calength,1);
    boxcal{i}.mag=ones(calength,1);
    boxcal{i}.phs=zeros(calength,1);
end
% reading box cals
fid=fopen([fpath fname],'r');
for i=1:5 
    % skipping some information (as we can read them from the TBL file)
    fgetl(fid);
end
for j=1:length(boxcal{1}.freq)
    if ~feof(fid)
        line=fgetl(fid);
        line=strrep(line,',',' ');
        if nc==5
            temp=sscanf(line,'%f %f %f %f %f %f %f %f %f %f %f');
        else
            temp=sscanf(line,'%f %f %f %f %f %f %f %f %f');
        end
        for i=1:nc          
            boxcal{i}.freq(j)=temp(1);
            boxcal{i}.mag(j)=temp(2*i);
            boxcal{i}.phs(j)=temp(2*i+1);
        end           
    end
end
% remove the trailing zeros (if any)
nrec = find(boxcal{1}.freq==0,1,"first");
for i=1:nc
    boxcal{i}.freq(nrec:end)=[];
    boxcal{i}.mag(nrec:end)=[];
    boxcal{i}.phs(nrec:end)=[];
end
fclose(fid);
return

File: src/read_boxcal.py
"""
read_boxcal - reads an ascii instrument response file of the legacy Phoenix
format (MTU-5A) and outputs channel-wise frequency response dictionaries.
"""

import os
from typing import Dict, List

import numpy as np


def _trim_padding(channel: Dict[str, np.ndarray]) -> Dict[str, np.ndarray]:
    first_zero = np.where(channel["freq"] == 0)[0]
    if first_zero.size == 0:
        return channel

    nrec = first_zero[0]
    channel["freq"] = channel["freq"][:nrec]
    channel["mag"] = channel["mag"][:nrec]
    channel["phs"] = channel["phs"][:nrec]
    return channel


def read_boxcal(fpath: str, fname: str, nc: int = 5) -> List[Dict[str, np.ndarray]]:
    """
    Parameters:
        fpath: path to the calibration file
        fname: name of calibration file (including extension)
        nc: number of channels (default 5 for MTU-5A)

    Returns:
        boxcal: list of channel dictionaries with keys:
                channel, freq, mag, phs
    """
    filepath = os.path.join(fpath, fname)
    if not os.path.isfile(filepath):
        raise FileNotFoundError(f"Box calibration {fname} not found")

    print(f"opening box calibration file: {fname}")

    calength = 150
    boxcal: List[Dict[str, np.ndarray]] = []
    for i in range(nc):
        boxcal.append(
            {
                "channel": i + 1,
                "freq": np.zeros(calength, dtype=np.float64),
                "mag": np.ones(calength, dtype=np.float64),
                "phs": np.zeros(calength, dtype=np.float64),
            }
        )

    with open(filepath, "r", encoding="latin-1") as fid:
        for _ in range(5):
            fid.readline()

        for j in range(calength):
            line = fid.readline()
            if not line:
                break

            line = line.replace(",", " ")
            vals = np.fromstring(line, sep=" ", dtype=np.float64)
            if vals.size == 0:
                continue

            expected = 11 if nc == 5 else 9
            if vals.size < expected:
                continue

            freq = vals[0]
            for i in range(nc):
                boxcal[i]["freq"][j] = freq
                boxcal[i]["mag"][j] = vals[2 * i + 1]
                boxcal[i]["phs"][j] = vals[2 * i + 2]

    for i in range(nc):
        boxcal[i] = _trim_padding(boxcal[i])

    return boxcal


if __name__ == "__main__":
    import sys

    if len(sys.argv) >= 3:
        cal = read_boxcal(sys.argv[1], sys.argv[2])
        print(f"Loaded {len(cal)} channels")
        if len(cal) > 0:
            print(f"Records in channel 1: {cal[0]['freq'].size}")
